From fb3e62bf2907ee596e3134f08a1e614a10bb16f6 Mon Sep 17 00:00:00 2001
From: Dmitry Muhomor <muhomor.dmitry@gmail.com>
Date: Mon, 27 Mar 2023 16:29:13 +0300
Subject: [PATCH 18/28] integrate Google's EuiccSupportPixel package

Depends on commit: "don't crash apps that depend on missing Gservices provider"
---
 data/etc/preinstalled-packages-platform.xml   |  4 +++
 .../server/ext/PackageManagerHooks.java       | 12 +++++++++
 .../server/pm/ext/EuiccSupportPixelHooks.java | 25 +++++++++++++++++++
 .../server/pm/ext/PackageHooksRegistry.java   |  2 ++
 4 files changed, 43 insertions(+)
 create mode 100644 services/core/java/com/android/server/pm/ext/EuiccSupportPixelHooks.java

diff --git a/data/etc/preinstalled-packages-platform.xml b/data/etc/preinstalled-packages-platform.xml
index e74a5edc1c13..6f92e88094b6 100644
--- a/data/etc/preinstalled-packages-platform.xml
+++ b/data/etc/preinstalled-packages-platform.xml
@@ -151,6 +151,10 @@ to pre-existing users, but cannot uninstall pre-existing system packages from pr
         <install-in user-type="android.os.usertype.profile.PRIVATE" />
     </install-in-user-type>
 
+    <install-in-user-type package="com.google.euiccpixel">
+        <install-in user-type="SYSTEM" />
+    </install-in-user-type>
+
    <!-- b/441173306 This is temporary fix to unblock the post submit failures
       @TODO vladokom@ to Remove this once the package is preinstalled in the
       correct XML to MK mapped configs
diff --git a/services/core/java/com/android/server/ext/PackageManagerHooks.java b/services/core/java/com/android/server/ext/PackageManagerHooks.java
index 9a750d4be31c..d3c8d73f0113 100644
--- a/services/core/java/com/android/server/ext/PackageManagerHooks.java
+++ b/services/core/java/com/android/server/ext/PackageManagerHooks.java
@@ -6,7 +6,9 @@ import android.app.AppBindArgs;
 import android.content.Context;
 import android.content.pm.ApplicationInfo;
 import android.content.pm.GosPackageState;
+import android.content.pm.PackageManager;
 import android.content.pm.PackageManagerInternal;
+import android.ext.PackageId;
 import android.os.Bundle;
 import android.os.UserHandle;
 import android.util.ArraySet;
@@ -27,6 +29,16 @@ public class PackageManagerHooks {
     @Nullable
     public static Integer maybeOverrideSystemPackageEnabledSetting(String pkgName, @UserIdInt int userId) {
         switch (pkgName) {
+            case PackageId.EUICC_SUPPORT_PIXEL_NAME:
+                if (userId == UserHandle.USER_SYSTEM) {
+                    // EuiccSupportPixel handles firmware updates of embedded secure element that is
+                    // used for eSIM, NFC, Felica, etc and should always be enabled.
+                    // It was previously unconditionally disabled after reboot.
+                    return PackageManager.COMPONENT_ENABLED_STATE_DEFAULT;
+                } else {
+                    // one of the previous OS versions enabled EuiccSupportPixel in all users
+                    return PackageManager.COMPONENT_ENABLED_STATE_DISABLED;
+                }
             default:
                 return null;
         }
diff --git a/services/core/java/com/android/server/pm/ext/EuiccSupportPixelHooks.java b/services/core/java/com/android/server/pm/ext/EuiccSupportPixelHooks.java
new file mode 100644
index 000000000000..6f44dd2ef122
--- /dev/null
+++ b/services/core/java/com/android/server/pm/ext/EuiccSupportPixelHooks.java
@@ -0,0 +1,25 @@
+package com.android.server.pm.ext;
+
+import android.Manifest;
+
+import com.android.internal.pm.pkg.parsing.PackageParsingHooks;
+import com.android.server.pm.pkg.PackageStateInternal;
+import com.android.internal.pm.pkg.component.ParsedUsesPermission;
+
+class EuiccSupportPixelHooks extends PackageHooks {
+
+    static class ParsingHooks extends PackageParsingHooks {
+        @Override
+        public boolean shouldSkipUsesPermission(ParsedUsesPermission p) {
+            // EuiccSupportPixel uses INTERNET perm only as part of its dev mode
+            return Manifest.permission.INTERNET.equals(p.getName());
+        }
+    }
+
+    @Override
+    public boolean shouldBlockPackageVisibility(int userId, PackageStateInternal otherPkg) {
+        // EuiccSupportPixel is a privileged package, block it from interacting with unprivileged
+        // parts of GMS
+        return isUserInstalledPkg(otherPkg);
+    }
+}
diff --git a/services/core/java/com/android/server/pm/ext/PackageHooksRegistry.java b/services/core/java/com/android/server/pm/ext/PackageHooksRegistry.java
index aca3bd31f42e..2e2198a96d56 100644
--- a/services/core/java/com/android/server/pm/ext/PackageHooksRegistry.java
+++ b/services/core/java/com/android/server/pm/ext/PackageHooksRegistry.java
@@ -8,12 +8,14 @@ public class PackageHooksRegistry {
 
     public static PackageParsingHooks getParsingHooks(String pkgName) {
         return switch (pkgName) {
+            case PackageId.EUICC_SUPPORT_PIXEL_NAME -> new EuiccSupportPixelHooks.ParsingHooks();
             default -> PackageParsingHooks.DEFAULT;
         };
     }
 
     public static PackageHooks getHooks(int packageId) {
         return switch (packageId) {
+            case PackageId.EUICC_SUPPORT_PIXEL -> new EuiccSupportPixelHooks();
             default -> PackageHooks.DEFAULT;
         };
     }
-- 
2.43.0

